### ISSUE TITLE
[Feature] Maintain relation field names from Prisma 1 datamodel 

### ISSUE THREAD

**Issue by nikolasburk** [prisma/upgrade#25]
The Upgrade CLI currently doesn't maintain the relation names from the Prisma 1 datamodel, for example this Prisma 1 datamodel:

```graphql
type User {
  id: ID! @id
  name: String!
  posts: [Post!]!
}

type Post {
  id: ID! @id
  title: String!
  author: User @relation(link: TABLE)
}
```

Gets turned into this Prisma schema:

```graphql
model Post {
  id        String     @id
  title     String
  User      User[]     @relation(references: [id])
}
 
model User {
  id       String   @id
  name     String
  Post     Post[]   @relation(references: [id])
}
```

Since the relation fields are part of the Prisma Client API surface, it would be very helpful if they could be named the same way as they were in the Prisma 1 datamodel since that means the user can recognize them immediately and isn't left wondering what this field is.

**Comment by matthewmueller** [comment]
A couple questions:

- Is this only for `link: TABLE`? 
- Is this an issue on many-to-many's & one-to-ones?

**Comment by nikolasburk** [comment]
I think this is a general issue with relation names, introspection will always output the model name (or even append some suffix if the relation needs to be disambiguate) but in Prisma 1, these relation fields all would have different/explicit names.

**Comment by nikolasburk** [comment]
## Examples

### 1-1 Relation

#### Prisma 1 datamodel

```graphql
type User {
  id: ID! @id
  name: String!
  profile: Profile
}

type Profile {
  id: ID! @id
  user: User! @relation(link: INLINE)
}
```

#### Current result from Upgrade CLI

```prisma
model User {
  id       String    @id
  Profile  Profile[]
}

model Profile {
  id   String  @id
  user String?
  User User?   @relation(fields: [user], references: [id])
}
```

#### Desired result from Upgrade CLI

```prisma
model User {
  id       String    @id
  Profile  Profile[]
}

model Profile {
  id     String  @id
  userId String? @map("user")
  user   User?   @relation(fields: [userId], references: [id])
}
```

---

### 1-n Relation

#### Prisma 1 datamodel

```graphql
type User {
  id: ID! @id
  name: String!
  posts: [Post!]!
}

type Post {
  id: ID! @id
  title: String!
  author: User @relation(link: TABLE)
}
```

#### Current result from Upgrade CLI

```prisma
model User {
  id       String    @id
  name     String
  Post     Post[]    @relation(references: [id])
}

model Post {
  id        String     @id
  title     String
  User      User[]     @relation(references: [id])
}
```

#### Desired result from Upgrade CLI

```prisma
model User {
  id       String    @id
  name     String
  posts    Post[]    @relation(references: [id])
}

model Post {
  id        String     @id
  title     String
  author    User[]     @relation(references: [id])
}
```

Note that this is still flawed because the `author` field is list instead of just a single field. This will be [treated separately by the Upgrade CLI](https://www.notion.so/prismaio/Limitations-of-Introspection-of-Prisma-1-schemas-Workarounds-24cfc8cb8cc3402281ee708f66339d42#21eeaef29947443eb5cb7a029a29bed8) (internal link) requiring a data migration by the user, but the CLI should use the previous relation name regardless. 

--- 

### m-n Relation

#### Prisma 1 datamodel

```graphql
type Post {
  id: ID! @id
  title: String!
  categories: [Category!]!
}

type Category {
  id: ID! @id
  name: String!
  posts: [Post!]!
}
```

#### Current result from Upgrade CLI

```prisma
model Post {
  id        String     @id
  title     String
  Category  Category[] @relation(references: [id])
}

model Category {
  id   String @id
  name String
  Post Post[] @relation(references: [id])
}
```

#### Desired result from Upgrade CLI

```prisma
model Post {
  id         String     @id
  title      String
  categories Category[] @relation(references: [id])
}

model Category {
  id    String @id
  name  String
  posts Post[] @relation(references: [id])
}
```

--- 

### Special cases

Prisma 1 didn't require to specify relations on both sides explicitly, meaning this was a valid Prisma 1 datamodel:

```graphql
type User {
  id: ID! @id
  name: String!
}

type Post {
  id: ID! @id
  title: String!
  author: User @relation(link: TABLE)
}
```

In this case, the Upgrade CLI should maintain only the one existing relation field name. For the second relation field (the one that's not specified in the Prisma 1 datamodel), it can use the one that was provide by introspection (that's just the name of the model). Ideally this could be lowercase though, so that all fields in the resulting Prisma 2 schema are lowercased. So this would be the desired result:

```diff
model User {
  id    String    @id
  name  String
+ post  Post[]    @relation(references: [id])
}

model Post {
  id        String     @id
  title     String
  author    User[]     @relation(references: [id])
}
```

**Comment by williamluke4** [comment]
@nikolasburk  Is there anything that I can do to speed this up?

**Comment by matthewmueller** [comment]
@williamluke4 since Prisma upgrade is meant to be a one-time tool, you can rename them manually in your schema to generate. 

It's still on my list. I hope to get to it this week, but I've been saying that for the last 2 weeks so let's see 😅

**Comment by williamluke4** [comment]
Thanks @matthewmueller, I just have a very large schema so I would highly appreciate this 😁

**Comment by matthewmueller** [comment]
This is implemented in `0.0.36`.

**Note** This is a 95% solution. It doesn't cover the following cases where there are multiple fields pointing back to the same model. 

```graphql
type User {
    bestFriend: User @relation(name: "UserBestFriend") # Optional self 1-1 relation
    godFather: User! @relation(name: "UserGodFather") # Required self 1-1 relation
    friends: [User!] @relation(name: "UserFriends", onDelete: CASCADE) # Self 1-m relation, cascade
}
```

In this case, you'll end up with duplicate field names

```prisma
model User {
  friendsId                User?                  @map("friends") @relation("UserToUser_bestFriend", fields: [bestFriend], references: [id])
  friendsId                User[]                 @map("friends") @relation("UserToUser_godFather")
}
```

We **do not plan** on fixing this, since it's a difficult problem. We recommend you install the [Prisma Language Tools](https://github.com/prisma/language-tools) for your IDE to detect these cases, then rename them manually.

You also may need to adjust some of the `@@index` field pointer to point from `user` to `userId` for example. Please follow the instructions that the Prisma formatter inside the language tools gives you. 

<img width="635" alt="CleanShot 2020-08-13 at 15 37 21@2x" src="https://user-images.githubusercontent.com/170299/90141149-06061800-dd7b-11ea-88b7-cef46bc404d4.png">

If you're still stuck, feel free to open an issue on this repo with your schema and we'll try our best to help you.
